tools/server-admin/src/main/java/org/projectnessie/tools/admin/cli/CutHistory.java

```diff
@@ -0,0 +1,128 @@
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.projectnessie.versioned.Hash;
+import org.projectnessie.versioned.ReferenceNotFoundException;
+import org.projectnessie.versioned.storage.common.logic.CommitLogic;
+import org.projectnessie.versioned.storage.common.logic.Logics;
+import org.projectnessie.versioned.storage.common.logic.PagedResult;
+import org.projectnessie.versioned.storage.common.objtypes.CommitObj;
+import org.projectnessie.versioned.storage.common.persist.Obj;
+import org.projectnessie.versioned.storage.common.persist.ObjId;
+import org.projectnessie.versioned.storage.versionstore.RefMapping;
+import picocli.CommandLine;
+
+@CommandLine.Command(
+    name = "cut-history",
+    mixinStandardHelpOptions = true,
+    description = {
+      "Advanced commit log manipulation command that detaches a number of most recent "
+          + "commits from their parent commit trail. Use with extreme caution. See also the 'cleanup-repository' "
+          + "command."
+    })
+public class CutHistory extends BaseCommand {
+  @CommandLine.Option(
+      names = {"-r", "--ref"},
+      description = "Reference name to use (default branch, if not set).")
+  private String ref;
+
+  @CommandLine.Option(
+      names = {"--depth", "-D"},
+      required = true,
+      description = "The number of commits to preserve intact starting from the ref HEAD.")
+  private int depth;
+
+  @Override
+  public Integer call() throws Exception {
+    if (!repositoryLogic(persist).repositoryExists()) {
+      spec.commandLine().getErr().println("Nessie repository does not exist");
+      return EXIT_CODE_REPO_DOES_NOT_EXIST;
+    }
+
+    CommitLogic commitLogic = Logics.commitLogic(persist);
+
+    Hash head = resolveRefHead();
+    List<CommitObj> toStore = new ArrayList<>();
+    Set<ObjId> danglingTail = new HashSet<>(); // commit IDs referenced from direct children
+    int count = depth;
+    PagedResult<CommitObj, ObjId> it = commitLogic.commitLog(commitLogQuery(hashToObjId(head)));
+    while (it.hasNext()) {
+      CommitObj commitObj = it.next();
+
+      if (count > 0) { // Processing a preserved commit
+        danglingTail.addAll(commitObj.tail());
+        danglingTail.remove(commitObj.id());
+        danglingTail.remove(EMPTY_OBJ_ID);
+
+        count--;
+        continue;
+      }
+
+      if (!commitObj.secondaryParents().isEmpty()) {
+        spec.commandLine()
+            .getErr()
+            .printf("ERROR: Unable to reset parents in merge commit '%s'%n", commitObj.id());
+        return EXIT_CODE_GENERIC_ERROR;
+      }
+
+      if (EMPTY_OBJ_ID.equals(commitObj.directParent())) {
+        spec.commandLine()
+            .getErr()
+            .printf(
+                "Encountered root commit '%s'. Full commit history is reserved.%n", commitObj.id());
+        return 0;
+      }
+
+      // Rewriting commit to fixup parents
```

Comment:
This is not enough. `CommitObj.tail()` contains (by default) up to 20 commit IDs, which means that more than one commit needs to be updated - otherwise all commit log related functionality will fail when trying to (pre)fetch no longer existing objects (in combination w/ #9753/#9688) during runtime.

Reply:
All commits referenced from the direct children's "tail" are rewritten (with the help of the `danglingTail` set).

I'll work on taking other references into account.